Add header with hash table class

LadyPython opened the pull request:
Semenova-Zvenigorodskaya Sofia 201-1
Solution https://official.contest.yandex.ru/contest/24924/run-report/52014905/

LadyPython committed:
Add header with hash table class

LadyPython committed:
Fix style

Semenova-Zvenigorodskaya Sofia 201-1
Solution https://official.contest.yandex.ru/contest/24924/run-report/51855699/

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,333 @@
+
+
+// Hash Table with Separate Chaining.
+// https://en.wikipedia.org/wiki/Hash_table#Separate_chaining
```
Вот тут недостаточно информации, всё-таки separate chaining не определяет стратегию перестроения при росте таблицы

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,333 @@
+    template<typename Iter>
+    HashMap(Iter start, Iter end, Hash hasher = Hash());
+
+    // Method returns the number of elements. O(1)
```
Хороший пример избыточного комментария, тут достаточно O(1)

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,333 @@
+
+    // Method inserts the element.
+    // If such key already exists, method does nothing.
+    // Amortized O(1). Double capacity if size == capacity.
```
увеличение capacity - это деталь реализации таблицы, не очень важно для публичного интерфейса метода insert

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,333 @@
+
+template<typename KeyType, typename ValueType, typename Hash>
+void HashMap<KeyType, ValueType, Hash>::insert(const ElementType& element) {
+    if (find(element.first) != end())
```
После if всегда должны быть фигурные скобки, даже если под условием одна строчка. Комментарий относится ко всей программе

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,333 @@
+}
+
+template<typename KeyType, typename ValueType, typename Hash>
+size_t HashMap<KeyType, ValueType, Hash>::rehash_key(const KeyType &key) const {
```
Казалось бы просто функция вычисляет hash от ключа и не совершает никаких перестроений, почему она называется rehash_key?

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,333 @@
+    auto new_capacity = 2 * sz_;
+    HashMapType new_hash_map(new_capacity);
+
+    for (auto hashed_key = 0ull; hashed_key < capacity_; ++hashed_key) {
```
почему тут auto, а не size_t?

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,333 @@
+    hash_map_[hashed_key].erase(it);
+    --sz_;
+
+    if (sz_ * 4 == capacity_)
```
константа 4 прямо в коде, параметры управляющие перестроением нужно вынести в отдельные константы или поля класса

glebshp replied:
Логику проверки необходимости провести rehash вынести в отдельный метод. В идеале сделать отдельный метод rehash_if_necessary, который будет вызывать проверку, и, если нужно, вызывать rehash

glebshp reviewed (changes requested):
Принципиальных проблем не вижу, нужны стилистические доработки

LadyPython committed:
Fix style

New solution: https://official.contest.yandex.ru/contest/24924/run-report/51878135/

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,345 @@
+template<typename KeyType, typename ValueType, typename Hash>
+template<typename Iter>
+HashMap<KeyType, ValueType, Hash>::HashMap(Iter start, Iter end, Hash hasher) : hasher_(hasher) {
+    for (auto it = start; it != end; ++it)
```
скобки

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,345 @@
+
+template<typename KeyType, typename ValueType, typename Hash>
+void HashMap<KeyType, ValueType, Hash>::erase(const KeyType& key) {
+    auto hashed_key = hash_key(key);
```
почему auto?

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,345 @@
+
+template<typename KeyType, typename ValueType, typename Hash>
+typename HashMap<KeyType, ValueType, Hash>::iterator HashMap<KeyType, ValueType, Hash>::find(const KeyType &key) {
+    auto hashed_key = hash_key(key);
```
тоже непонятно, почему auto

LadyPython committed:
Fix style

Remove `auto` from not iterators and add force braces.

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,348 @@
+
+template<typename KeyType, typename ValueType, typename Hash>
+void HashMap<KeyType, ValueType, Hash>::rehash_if_necessary() {
+    if (sz_ * INCREASING_SIZE_COEFFICIENT != capacity_ && sz_ * DECREASING_SIZE_COEFFICIENT != capacity_) {
```
мне кажется, что такая строгая проверка на != вместо > и < сейчас работает только из-за того что все константы являются степенями двойки, если там поставить 7 и 3, то масштабирование сломается

LadyPython committed:
Fix bug with rehashing
